core/src/main/java/io/opencensus/stats/Measure.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2016, Google Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.stats;
+
+import com.google.auto.value.AutoValue;
+import io.opencensus.common.Function;
+import io.opencensus.internal.StringUtil;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * Measure.
+ *
+ * <p>Note: Measure names are {@link String}s with enforced restrictions.
+ */
+@Immutable
+public abstract class Measure {
+
+  public static final int MAX_LENGTH = StringUtil.MAX_LENGTH;
+
+  /**
+   * Applies the given match function to the underlying data type.
+   */
+  public abstract <T> T match(Function<DoubleMeasure, T> p0, Function<LongMeasure, T> p1);
+
+  /**
+   * Name of measure, e.g. rpc_latency, cpu. Must be unique.
+   */
+  public abstract Measure.Name getMeasureName();
+
+  /**
+   * Name of measure, as a {@code String}.
+   */
+  public final String getName() {
+    return getMeasureName().asString();
+  }
+
+  /**
+   * Detailed description of the measure, used in documentation.
+   */
+  public abstract String getDescription();
+
+  /**
+   * The units in which {@link Measure} values are measured.
+   *
+   * <p>The grammar for a unit is as follows:
+   *     Expression = Component { "." Component } { "/" Component } ;
+   *     Component = [ PREFIX ] UNIT [ Annotation ] | Annotation | "1" ;
+   *     Annotation = "{" NAME "}" ;
+   * For example, string “MBy{transmitted}/ms” stands for megabytes per milliseconds, and the
+   * annotation transmitted inside {} is just a comment of the unit.
+   */
+  public abstract String getUnit();
+
+  /**
+   * The name of a {@code Measure}.
+   */
+  // This type should be used as the key when associating data with Measures.
+  @Immutable
+  @AutoValue
+  public abstract static class Name {
```

Comment:
I think if we would allow different types of `Measure` to have the same name, we'll need to have typed subclasses on `Name`, but this is not determined yet. Anyway for now I'll go ahead and replace the `Name` class with plain string, we might add it back later if it's necessary.